pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,244 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import IterableDataset
+
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler(Sampler):
+    """
+    This FastForwardSampler wraps a :class:`torch.utils.data.Sampler` and records the number of iterations
+    performed during an epoch. It maintains a state through :meth:`state_dict` that can be reloaded through
+    :meth:`load_state_dict`. If the sampler is used in a multiprocessing context, the FastForwardSampler will record
+    the state of the current worker.
+
+    When reloading, the FastForwardSampler will `fast-forward` the wrapped sampler by iterating through all the
+    samples seen in the last iterations (for the current worker).
+    """
+
+    def __init__(self, sampler: Union[Sampler, BatchSampler, Generator]) -> None:
+        super().__init__(data_source=None)
+        self.sampler = sampler
+        self.restarting: bool = False
+        self._current_iteration = 0
+        self._dataloader_batch_size: Optional[int] = None
+        self._cached_state_dict: Optional[Dict[str, Any]] = None
+
+    def __getattr__(self, key: str) -> Any:
+        if key in self.__dict__:
+            return self.__dict__[key]
+        return self.sampler.__dict__[key]
+
+    def setup(self, dataloader_batch_size: Optional[int] = None) -> None:
+        """
+        Setup the ``FastForwardSampler``.
+        This is required only when the provided dataset subclassed :class:`torch.utils.data.Dataset`.
+        """
+        self._dataloader_batch_size = dataloader_batch_size
+
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    @property
+    def drop_last(self) -> bool:
+        return self.sampler.drop_last
+
+    @property
+    def batch_size(self) -> int:
+        return self.sampler.batch_size
+
+    @property
+    def batch_indices(self) -> Optional[List[int]]:
+        return self.sampler.batch_indices
+
+    def __iter__(self) -> Iterator[List[int]]:
```

Comment:
Technically, this could be `Any` for some weird samplers?

```suggestion
    def __iter__(self) -> Iterator[Any]:
```

Reply:
Done !